lib/AST/SubstitutionMap.cpp

```diff
@@ -64,173 +72,115 @@ Type SubstitutionMap::lookupSubstitution(CanSubstitutableType type) const {
 
 void SubstitutionMap::
 addSubstitution(CanSubstitutableType type, Type replacement) {
+  assert(!(type->isTypeParameter() && !getGenericSignature()) &&
+         "type parameter substitution map without generic signature");
+  assert(!(type->hasArchetype() &&
+           !genericSigOrEnv.is<GenericEnvironment *>()) &&
+         "archetype substitution map without generic environment");
   auto result = subMap.insert(std::make_pair(type, replacement));
   assert(result.second || result.first->second->isEqual(replacement));
   (void) result;
 }
 
-template<typename T>
-Optional<T> SubstitutionMap::forEachParent(
-              CanType type,
-              llvm::SmallPtrSetImpl<CanType> &visitedParents,
-              llvm::function_ref<Optional<T>(CanType,
-                                             AssociatedTypeDecl *)> fn) const {
-  // If we've already visited the parents of this type, stop.
-  if (!visitedParents.insert(type).second)
-    return None;
-
-  auto foundParents = parentMap.find(type.getPointer());
-  if (foundParents != parentMap.end()) {
-    for (auto parent : foundParents->second) {
-      if (auto result = fn(parent.first, parent.second))
-        return result;
-    }
+Optional<ProtocolConformanceRef>
+SubstitutionMap::lookupConformance(CanType type, ProtocolDecl *proto) const {
+  // If we have an archetype, map out of the context so we can compute a
+  // conformance access path.
+  GenericEnvironment *genericEnv = nullptr;
+  if (auto archetype = type->getAs<ArchetypeType>()) {
+    genericEnv = archetype->getGenericEnvironment();
+    type = genericEnv->mapTypeOutOfContext(type)->getCanonicalType();
   }
 
-  if (auto archetypeType = dyn_cast<ArchetypeType>(type))
-    if (auto *parent = archetypeType->getParent())
-      return fn(CanType(parent), archetypeType->getAssocType());
-
-  if (auto memberType = dyn_cast<DependentMemberType>(type))
-    return fn(CanType(memberType->getBase()), memberType->getAssocType());
+  // Error path: if we don't have a type parameter, there is no conformance.
+  // FIXME: Query concrete conformances in the generic signature?
+  if (!type->isTypeParameter())
+    return None;
 
-  return None;
-}
+  // Retrieve the starting conformance from the conformance map.
+  auto getInitialConformance =
+    [&](Type type, ProtocolDecl *proto) -> Optional<ProtocolConformanceRef> {
+      // We're working relative to a generic environment, map into that
+      // context before looking into the conformance map.
+      if (genericEnv)
```

Comment:
It seems silly to map the type out of context only to map it back in again. What do you think of always using interface types as keys in the conformance map, even for generic environment-derived substitutions?